presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/PhysicalCteOptimizer.java

```diff
@@ -0,0 +1,259 @@
+import com.facebook.presto.metadata.PartitioningMetadata;
+import com.facebook.presto.spi.ColumnHandle;
+import com.facebook.presto.spi.ColumnMetadata;
+import com.facebook.presto.spi.PrestoException;
+import com.facebook.presto.spi.TableHandle;
+import com.facebook.presto.spi.VariableAllocator;
+import com.facebook.presto.spi.WarningCollector;
+import com.facebook.presto.spi.plan.Assignments;
+import com.facebook.presto.spi.plan.CteConsumerNode;
+import com.facebook.presto.spi.plan.CteProducerNode;
+import com.facebook.presto.spi.plan.PlanNode;
+import com.facebook.presto.spi.plan.PlanNodeIdAllocator;
+import com.facebook.presto.spi.plan.ProjectNode;
+import com.facebook.presto.spi.plan.TableScanNode;
+import com.facebook.presto.spi.relation.VariableReferenceExpression;
+import com.facebook.presto.sql.planner.BasePlanFragmenter;
+import com.facebook.presto.sql.planner.Partitioning;
+import com.facebook.presto.sql.planner.TypeProvider;
+import com.facebook.presto.sql.planner.plan.SimplePlanRewriter;
+import com.google.common.collect.ImmutableList;
+
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+
+import static com.facebook.presto.SystemSessionProperties.getHashPartitionCount;
+import static com.facebook.presto.SystemSessionProperties.isMaterializeAllCtes;
+import static com.facebook.presto.spi.StandardErrorCode.NOT_SUPPORTED;
+import static com.facebook.presto.sql.TemporaryTableUtil.assignPartitioningVariables;
+import static com.facebook.presto.sql.TemporaryTableUtil.assignTemporaryTableColumnNames;
+import static com.facebook.presto.sql.TemporaryTableUtil.createTemporaryTableScan;
+import static com.facebook.presto.sql.TemporaryTableUtil.createTemporaryTableWriteWithoutExchanges;
+import static com.google.common.collect.ImmutableList.toImmutableList;
+import static java.lang.String.format;
+
+public class PhysicalCteOptimizer
+        implements PlanOptimizer
+{
+    private final Metadata metadata;
+
+    public PhysicalCteOptimizer(Metadata metadata)
+    {
+        this.metadata = metadata;
+    }
+
+    @Override
+    public PlanOptimizerResult optimize(PlanNode plan, Session session, TypeProvider types, VariableAllocator variableAllocator, PlanNodeIdAllocator idAllocator, WarningCollector warningCollector)
+    {
+        if (!isMaterializeAllCtes(session)) {
+            return PlanOptimizerResult.optimizerResult(plan, false);
+        }
+        PhysicalCteTransformerContext context = new PhysicalCteTransformerContext();
+        PlanNode producerReplaced = SimplePlanRewriter.rewriteWith(new CteProducerRewiter(session, idAllocator, variableAllocator), plan, context);
+        PlanNode rewrittenPlan = SimplePlanRewriter.rewriteWith(new CteConsumerRewrite(session, idAllocator, variableAllocator), producerReplaced, context);
+        return PlanOptimizerResult.optimizerResult(rewrittenPlan, !rewrittenPlan.equals(plan));
```

Comment:
`rewrittenPlan.equals(plan)` may not be a good way, which involves false positives

Reply:
In this case, there are no rewrites if no cte producers or consumers are present and can be used